pytorch_toolkit/formula_recognition/tools/test.py

```diff
@@ -0,0 +1,98 @@
+                                                Im2LatexDataset)
+from im2latex.models.im2latex_model import Im2latexModel
+from torch.utils.data import DataLoader
+from tools.evaluation_tools import Im2latexRenderBasedMetric
+
+
+class Evaluator():
+    def __init__(self, config):
+        self.config = config
+        self.model_path = config.get('model_path')
+        self.val_path = config.get('val_path')
+        self.vocab = read_vocab(config.get('vocab_path'))
+        self.val_transforms_list = config.get('val_transforms_list')
+        self.split = config.get('split_file', 'validate')
+        self.print_freq = config.get('print_freq', 16)
+        self.load_dataset()
+        self.model = Im2latexModel(config.get('backbone_type'), config.get(
+            'backbone_config'), len(self.vocab), config.get('head'))
+        self.device = config.get('device', 'cpu')
+        if self.model_path is not None:
+            self.model.load_weights(self.model_path, old_model=config.get("old_model"), map_location=self.device)
+
+        self.model = self.model.to(self.device)
+        self.time = get_timestamp()
+
+    def load_dataset(self):
+
+        val_dataset = Im2LatexDataset(self.val_path, self.split)
+        val_sampler = BatchRandomSampler(dataset=val_dataset, batch_size=1)
+        batch_transform_val = create_list_of_transforms(self.val_transforms_list)
+        self.val_loader = DataLoader(
+            val_dataset,
+            batch_sampler=val_sampler,
+            collate_fn=partial(collate_fn, self.vocab.sign2id,
+                               batch_transform=batch_transform_val),
+            num_workers=os.cpu_count())
+
+    def validate(self):
+        self.model.eval()
+        print("Validation started")
+        annotations = []
+        predictions = []
+        metric = Im2latexRenderBasedMetric()
+        with torch.no_grad():
+
+            for img_name, imgs, tgt4training, tgt4cal_loss in tqdm(self.val_loader, initial=1):
+                imgs = imgs.to(self.device)
+                tgt4training = tgt4training.to(self.device)
+                tgt4cal_loss = tgt4cal_loss.to(self.device)
+                _, pred = self.model(imgs)
+                gold_phrase_str = self.vocab.construct_phrase(tgt4cal_loss[0])
+                pred_phrase_str = self.vocab.construct_phrase(pred[0], max_len=1 + len(gold_phrase_str.split()))
+
+                annotations.append((gold_phrase_str, img_name[0]))
+                predictions.append((pred_phrase_str, img_name[0]))
+        res = metric.evaluate(annotations, predictions)
+        return res
+
+
+def parse_args():
+    args = argparse.ArgumentParser()
+    args.add_argument('--config')
+    return args.parse_args()
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    with open(args.config, 'r') as f:
+        config = yaml.load(f, Loader=yaml.SafeLoader)
+    validator = Evaluator(config)
+    print("Im2latex metric is: {}".format(validator.validate()))
```

Comment:
Please, use a separate variable to get the result of validation (e.g. `result_validation`).   
Do not hide the main operation in the script -- you shouldn't be ashamed of it!

Reply:
Done